capstone/src/test/java/com/google/sps/servlets/student/StudentServletTest.java

```diff
@@ -0,0 +1,177 @@
+    StringWriter stringWriter = new StringWriter();
+    PrintWriter printWriter = new PrintWriter(stringWriter);
+    when(response.getWriter()).thenReturn(printWriter);
+    studentServlet.doGet(request, response);
+
+    String responseStr = stringWriter.getBuffer().toString().trim();
+    JsonElement responseJsonElement = new JsonParser().parse(responseStr);
+    JsonObject responseJson = responseJsonElement.getAsJsonObject();
+
+    return responseJson;
+  }
+
+  private String doPostStudentServletResponse() throws ServletException, IOException {
+    StringWriter stringWriter = new StringWriter();
+    PrintWriter printWriter = new PrintWriter(stringWriter);
+    when(response.getWriter()).thenReturn(printWriter);
+
+    studentServlet.doPost(request, response);
+    return stringWriter.toString();
+  }
+
+  @Test
+  public void doGet_studentLogsInForFirstTime() throws ServletException, IOException {
+    localHelper.setEnvEmail(KEVIN_EMAIL).setEnvAuthDomain("google.com").setEnvIsLoggedIn(true);
+    when(request.getParameter(Constants.PROPERTY_EMAIL)).thenReturn(KEVIN_EMAIL);
+
+    JsonObject responseJson = doGetStudentServletResponse();
+    JsonObject responseStudent = responseJson.get(STUDENT).getAsJsonObject();
+
+    String responseName = responseStudent.get(Constants.PROPERTY_NAME).getAsString();
+    String responseEmail = responseStudent.get(Constants.PROPERTY_EMAIL).getAsString();
+    String responseMajor = responseStudent.get(Constants.PROPERTY_MAJOR).getAsString();
+    int responseYear = responseStudent.get(Constants.PROPERTY_GRADYEAR).getAsInt();
+    ImmutableList responseClubs =
+        ImmutableList.copyOf(responseStudent.get(Constants.PROPERTY_CLUBS).getAsJsonArray());
+
+    Assert.assertEquals("First Last", responseName);
+    Assert.assertEquals(KEVIN_EMAIL, responseEmail);
+    Assert.assertEquals(0, responseYear);
+    Assert.assertEquals("", responseMajor);
+    Assert.assertEquals(ImmutableList.of(), responseClubs);
+  }
+
+  @Test
+  public void doGet_studentIsInNoClubs() throws ServletException, IOException {
+    localHelper.setEnvEmail(MEGHA_EMAIL).setEnvAuthDomain("google.com").setEnvIsLoggedIn(true);
+    when(request.getParameter(Constants.PROPERTY_EMAIL)).thenReturn(MEGHA_EMAIL);
+
+    JsonObject responseJson = doGetStudentServletResponse();
+    JsonObject responseStudent = responseJson.get(STUDENT).getAsJsonObject();
+
+    String responseName = responseStudent.get(Constants.PROPERTY_NAME).getAsString();
+    String responseEmail = responseStudent.get(Constants.PROPERTY_EMAIL).getAsString();
+    String responseMajor = responseStudent.get(Constants.PROPERTY_MAJOR).getAsString();
+    int responseYear = responseStudent.get(Constants.PROPERTY_GRADYEAR).getAsInt();
+    ImmutableList responseClubs =
+        ImmutableList.copyOf(responseStudent.get(Constants.PROPERTY_CLUBS).getAsJsonArray());
+
+    Assert.assertEquals(student_megha.getProperty(Constants.PROPERTY_NAME), responseName);
```

Comment:
So I guess this is what I was referring to about the constants. student_megha is defined globally but really only used within the scope of this test, same with student_megan (and even if it were used in multiple tests, it's ok to define it globally, but you should only add it to the Datastore in the relevant tests).

I'm unfamiliar with Datastore and how the setup works, but is it possible to only define and add the relevant entities to Datastore within the relevant tests (not in setUp). This would also allow you to eliminate all of the constants as you can just have those while defining the entity within the test. I _think_ every test does its own setUp and tearDown so adding an entity to Datastore in one test won't affect the others. If that doesn't work, maybe you can clear Datastore manually (either in tearDown or at the beginning of each individual test). This way, you explicitly define and control the test behavior and expectations within each test. 

I think this should work, but if it doesn't feel free to submit this as-is.

Reply:
It worked! Now I define the entity and put the entity in Datastore for each test case instead of in the setUp() method